### PR TITLE
Bind VCMI server to a random TCP port

### PR DESCRIPTION
**Pull request by smanolloff**
### Overview

This PR allows to set `settings["server"]["localPort"]=0` in order to bind VCMI server to a random TCP port.

### Problem statement

For the purposes of ML, being able to run multiple VCMI instances in parallel is essential in order to efficiently use the available hardware resources by training multiple AI agents simultaneously. Since there can only be one process bound to a given port, the workaround is to manually select a free TCP port prior to launching the application, but there are many issues with this approach.

In addition, when running VCMI server+client locally, there is a (small) chance of a port conflict with third party software using the same port.

### Changes summary

* `settings["server"]["localPort"]` can now be set to `0` (i.e. randomly assigned port)
* A new function `CVCMIServer::prepare` (naming suggestions are welcome) must now be called in order to bind to a network socket (this used to be part of the server constructor). This function returns the assigned port.
* Several functions (`IServerRunner::start`, `ServerProcessRunner::start`, `ServerThreadRunner::start`, `NetworkInterface::start`, etc.) return type was changed from `void` to `uint16_t` (the assigned port).

### Example

Using a `localPort` of `0`, we can see that a random port is assigned (63615 in this case):

```
[0x1eaadfac0][network] INFO Randomly assigned port will be used
[0x1eaadfac0][network] INFO Listening for connections at port 63615
```

Subsequently started VCMI server processes will bind to a different port without conflict.

### CODE REVIEW

File: client/ServerRunner.cpp [hunk 1]
```diff
@@ -23,19 +23,30 @@
...
 		setThreadName("runServer");
+		srvport = server->prepare(connectToLobby);
 		server->run();
 	});
+
+	while(srvport == 0) {
+		logNetwork->trace("Waiting for server port...");
+		boost::this_thread::sleep(boost::posix_time::milliseconds(100));
+		logNetwork->debug("Server port: %d", srvport);
+	}
```
Comment (IvanSavenko) [inline client/ServerRunner.cpp:47]: Can we do something better than sleep-based thread synchronization? std::future for example? Or anything else that is not sleep?

File: lib/network/NetworkServer.cpp [hunk 2]
```diff
@@ -19,16 +19,17 @@ NetworkServer::NetworkServer(INetworkServerListener & listener, const std::share
 {
 }
 
-void NetworkServer::start(uint16_t port)
+uint16_t NetworkServer::start(uint16_t port)
 {
-	acceptor = std::make_shared<NetworkAcceptor>(*io, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), port));
-	startAsyncAccept();
+	acceptor = std::make_shared<NetworkAcceptor>(*io, boost::asio::ip::tcp::endpoint(boost::asio::ip::address::from_string("127.0.0.1"), port));
```
Comment (IvanSavenko) [inline lib/network/NetworkServer.cpp:24]: Why was this change necessary?

File: client/ServerRunner.cpp (hunk 1 above)
Comment (smanolloff) [reply client/ServerRunner.cpp:47]: Sure, I will look into that.

File: lib/network/NetworkServer.cpp (hunk 2 above)
Comment (smanolloff) [reply lib/network/NetworkServer.cpp:24]: oops. Not only is it not necessary, it is a mistake to include it in the PR. I will remove it, thanks for being vigilant.

Context: VCMI binds to 0.0.0.0 by default, which produces an annoying OS firewall warning and that's why I had changed it to localhost. That would, of course, prevent external connections to the server, so, yeah, I will change it back.

**Comment by Laserlicht** [comment]
@smanolloff Thanks for your contribution! Do you plan to backport more of your gym environment stuff in upstream? Would be great if we had an easy to use environment to train models (that will be keep in sync with current VCMI version). And maybe in the future a good neural network based AI.

Feel free to also join discord. There are some discussion about weak points of current AI.

**Comment by smanolloff** [comment]
> @smanolloff Thanks for your contribution! Do you plan to backport more of your gym environment stuff in upstream? Would be great if we had an easy to use environment to train models (that will be keep in sync with current VCMI version). And maybe in the future a good neural network based AI.
> 
> Feel free to also join discord. There are some discussion about weak points of current AI.

Oh, yes, actually I am finalizing some of the changes there and was wondering what would be the proper approach for such a contribution. I even have some decent AI models trained already which seem to outperform the BattleAI bot. I am currently working on abstracting away the changes in order to make them as less intrusive as possible, but, yeah, a discussion in the Discord channel is surely a good idea, I am planning to approach you (and the other VCMI devs) there soon to work out the possible rollout strategies for such a feature :D And thank you 🍺 Cheers

**Comment by Laserlicht** [comment]
Would it be possible to make this default. Any drawbacks?

For testing I've often multiple instances open. Which crashes with bind. Even on android if you have develop version in background and try to start release version.

**Comment by Laserlicht** [comment]
Maybe only if default port fails.
